moveit_experimental/jog_arm/src/jog_arm/jog_calcs.cpp

```diff
@@ -0,0 +1,763 @@
+  delta_x[4] = vector_toward_singularity[4] / scale;
+  delta_x[5] = vector_toward_singularity[5] / scale;
+
+  // Calculate a small change in joints
+  Eigen::VectorXd new_delta_theta = pseudo_inverse_ * delta_x;
+
+  double theta[6];
+  const double* prev_joints = kinematic_state_->getVariablePositions();
+  for (std::size_t i = 0, size = static_cast<std::size_t>(num_joints_); i < size; ++i)
+    theta[i] = prev_joints[i] + new_delta_theta(i);
+
+  kinematic_state_->setVariablePositions(theta);
+  jacobian_ = kinematic_state_->getJacobian(joint_model_group_);
+  Eigen::JacobiSVD<Eigen::MatrixXd> new_svd = Eigen::JacobiSVD<Eigen::MatrixXd>(jacobian_);
+  double new_condition = new_svd.singularValues()(0) / new_svd.singularValues()(new_svd.singularValues().size() - 1);
+  // If new_condition < ini_condition, the singular vector does point towards a
+  // singularity. Otherwise, flip its direction.
+  if (ini_condition >= new_condition)
+  {
+    vector_toward_singularity[0] *= -1;
+    vector_toward_singularity[1] *= -1;
+    vector_toward_singularity[2] *= -1;
+    vector_toward_singularity[3] *= -1;
+    vector_toward_singularity[4] *= -1;
+    vector_toward_singularity[5] *= -1;
+  }
+
+  // If this dot product is positive, we're moving toward singularity ==> decelerate
+  double dot = vector_toward_singularity.dot(commanded_velocity);
+  if (dot > 0)
+  {
+    // Ramp velocity down linearly when the Jacobian condition is between lower_singularity_threshold and
+    // hard_stop_singularity_threshold, and we're moving towards the singularity
+    if ((ini_condition > parameters_.lower_singularity_threshold) &&
+        (ini_condition < parameters_.hard_stop_singularity_threshold))
+    {
+      velocity_scale = 1. -
+                       (ini_condition - parameters_.lower_singularity_threshold) /
+                           (parameters_.hard_stop_singularity_threshold - parameters_.lower_singularity_threshold);
+    }
+
+    // Very close to singularity, so halt.
+    else if (ini_condition > parameters_.hard_stop_singularity_threshold)
+    {
+      velocity_scale = 0;
+      ROS_WARN_THROTTLE_NAMED(2, LOGNAME, "Close to a singularity. Halting.");
+    }
+  }
+
+  return velocity_scale;
+}
+
+void JogCalcs::enforceJointVelocityLimits(Eigen::VectorXd& calculated_joint_velocity)
+{
+  double maximum_joint_vel = calculated_joint_velocity.cwiseAbs().maxCoeff();
+  if (maximum_joint_vel > parameters_.joint_scale)
+  {
+    // Scale the entire joint velocity vector so that each joint velocity is below min, and the output movement is
+    // scaled uniformly to match expected motion
+    calculated_joint_velocity = calculated_joint_velocity * parameters_.joint_scale / maximum_joint_vel;
+  }
+}
+
+bool JogCalcs::checkIfJointsWithinBounds(trajectory_msgs::JointTrajectory& new_jt_traj)
+{
+  bool halting = false;
+  for (auto joint : joint_model_group_->getJointModels())
+  {
+    if (!kinematic_state_->satisfiesVelocityBounds(joint))
```

Comment:
Is kinematics_state the actual velocity? Why would you set the joint velocity to the maximum velocity, if you've already checked the calculated set value in `JogCalcs::enforceJointVelocityLimits`? Why is this duplication necessary?

Reply:
It took me awhile to recall, but the answer is:
enforceVelocityLimits() ensures the velocity matches the limits in the jog_arm config file.

kinematic_state_ checks the URDF limits.

Will change the function name to clarify - thanks!